File: parsers/find-words-in-dump.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#define BEFORE 10

int found(const char *context,const char *text) {
   int l = strlen(text);
   return strncmp(text, context+BEFORE-l, l) == 0;
}

int main()
{
   int milon_count=0;
   int i,j,k=0,lll=0,existing=0,textflag=0,fileflag=1,imageflag=1,templateflag=1,italicflag=1,divflag=1,
   codeflag=1,sourceflag=1,doublenewline=1,poemflag=1,
   galleryflag=1,linkflag=1,quoteflag=1,doublequoteflag=1,doublequoteflag2=1,columnsflag=1,
   spaceflag=0,noteflag=1,noteflag2=1,noteflag3=1,namecount=0,nameflag=0,namespaceflag=0, wiktionflag=1;
   int saweng=0;
   char typo[80];
   char oldtypo[80];
   char pagename[100];
   char *correction;
   char *method;
   char *location;
   char lastchar=0,lastchar2;
   char (*milon)[61];
   char context[BEFORE];
   milon = calloc (22000000, sizeof(char[61]));

   // to access character i of word w
   FILE *fp1 = fopen("enwiki-20220701-pages-articles.xml", "r");
   FILE *fp2 = fopen("typos.txt", "r");
   FILE *fp3 = fopen("file3.txt", "w");

   if (fp1 == NULL || fp2 == NULL || fp3 == NULL )
   {
   puts("Could not open files");
   exit(0);
   }

   j=0;
   k=0;
   char c;
   while ((c = fgetc(fp2)) != EOF) //load main milon
   {
      milon[j][k]=c;
      if(j==22000000)break;
      if(k>59&&c!='\n')continue;
      if(c==',')milon[j][k]=0;
      if(c=='\n'){milon[j][k]=0;milon_count++;j++;k=0;} //end of word - move to next word
      if (c!='\n')k++;
   }

   j=0;
   c=0;
   while (c!=EOF){//word loop
      k=0;typo[0]=0;
   while ((c = fgetc(fp1))!=EOF){//letter loop

      //save article name: if <title>
      if(found(context,"<title>")){
         nameflag=1;
      }

      if(nameflag==1){pagename[namecount]=c;namecount++;}
      if(nameflag==1&&context[BEFORE-1]=='<'&&c=='/'){pagename[namecount-2]=0;namecount=0;nameflag=0;}

      //close tags in new article
      if(nameflag==1)
         {divflag=1;codeflag=1;galleryflag=1;templateflag=1;doublenewline=1;
                  columnsflag=1;sourceflag=1;noteflag=1;noteflag2=1;noteflag3=1;}
      //only search within articles text
      if(found(context,"<text")){textflag=1;}
      if(found(context,"</text")){typo[0]=0;textflag=0;}
      //only search namespace 0
      if(found(context,"<ns>")){
         if(c=='0'){namespaceflag=1;}
         else{typo[0]=0;namespaceflag=0;}
      }
      //skip files
      if(found(context,"File") || found(context,"file")){typo[0]=0;fileflag=0;}
      if(fileflag==0&&c=='\n'){fileflag=1;}
      if(found(context,"Image") || found(context,"image")){typo[0]=0;imageflag=0;}
      if(imageflag==0&&c=='\n'){imageflag=1;}
      //skip templates
      if(found(context,"{{")){
         typo[0]=0;
         doublenewline=0;
         templateflag++;
      }
      if(context[BEFORE-1]=='}'&&c!='}'){
         templateflag--;
      }
      if(found(context,"\n\n")){doublenewline=1;}
      //skip links
      if(found(context,"[h")){typo[0]=0;linkflag=0;}
      if(linkflag==0&&c==']'){linkflag=1;}
      if(linkflag==0&&c=='\n'){linkflag=1;}
      //skip blockquotes
      if(found(context,";bloc")){typo[0]=0;quoteflag=0;}
      if(found(context,";/bloc")){typo[0]=0;quoteflag=1;}
      if(nameflag==1){quoteflag=1;}
      //skip doublequotes
      if(context[BEFORE-1]=='"'){typo[0]=0;doublequoteflag=0;}
      if(doublequoteflag==0&&c=='"'){doublequoteflag=1;}
      //skip doublequotes2
      if(found(context,"&quot;")){typo[0]=0;doublequoteflag2=0;}
      if(found(context,"&quot")){doublequoteflag2=1;}
      if(doublequoteflag2==0&&c=='\n'){doublequoteflag2=1;}
      //skip italics
      if(found(context,"\'\'")){typo[0]=0;italicflag=0;}
      if(context[BEFORE-1]=='\''&&c=='\''){italicflag=1;}
      if(italicflag==0&&c=='\n'){italicflag=1;}
      //skip columns
      if(found(context,"{|")){typo[0]=0;columnsflag=0;}
      if(found(context,"|}")){columnsflag=1;}
      //skip notes
      if(found(context,";ref&gt;") || found(context,";ref &gt;")){typo[0]=0;noteflag=0;}
      if(found(context,";/ref&")){noteflag=1;}
      //skip notes
      if(found(context,";!--")){typo[0]=0;noteflag2=0;}
      if(found(context,"&gt;")){noteflag2=1;}
      //skip notes
      if(found(context,";ref name")){typo[0]=0;noteflag3=0;}
      if(found(context,"--&gt;")){noteflag3=1;}
      //skip div
      if(found(context,";div")){typo[0]=0;divflag=0;}
      if(found(context,";/div&")){divflag=1;}
      //skip <code>
      if(found(context,";code")){typo[0]=0;codeflag=0;}
      if(found(context,";/code")){codeflag=1;}
      //skip <poem>
      if(found(context,";poem")){typo[0]=0;poemflag=0;}
      if(found(context,";/poem")){poemflag=1;}
      //skip <gallery>
      if(found(context,";gallery")){typo[0]=0;galleryflag=0;}
      if(found(context,";/gallery")){galleryflag=1;}     
      //skip <source>
      if(found(context,";source")){typo[0]=0;sourceflag=0;}
      if(found(context,";/source")){sourceflag=1;}

      //save context
      for(i=0;i<BEFORE;i++){context[i]=context[i+1];}
      context[BEFORE-1]=c;

      if(k==79){spaceflag=0;break;}

      if(namespaceflag==1&&textflag==1&&spaceflag==1&&fileflag==1&&imageflag==1&&
         galleryflag==1&&italicflag==1&&columnsflag==1&&sourceflag==1&&doublenewline==1&&
         templateflag==1&&noteflag==1&&noteflag2==1&&noteflag3==1&&divflag==1&&codeflag==1&&
         linkflag==1&&quoteflag==1&&doublequoteflag==1&&doublequoteflag2==1&&wiktionflag==1)
         {typo[k]=c;k++;}

      //ignore short links
      if(typo[k-2]=='.'&&typo[k-1]!=' '&&typo[k-1]!='\n')
         {typo[0]='\0';spaceflag=0;break;}

      //chars to separate words and initialize space
      if (c=='\n'||c==' '||c=='\t')
         {typo[k-1]='\0';spaceflag=1;break;}

      //chars to separate words and check
      if (c=='-'||c=='}'||c=='['||c==','||c=='.'||c=='"'||c==';'||
         c=='^'||c=='~'||c=='#'||c=='='||c=='&'||c==')'||
         c=='|'||c=='<'||c=='>'||c=='?'||c=='^'||c=='_')
      {typo[k-1]='\0';break;}

      //chars to delete words and wait for space
      if(k>0&&(c==':'||c=='%'||c=='('||c=='/'||c=='\\'||c=='@'||c=='$'||c=='!'||c=='{'))
      {typo[0]='\0';spaceflag=0;break;}

   }
      if(strlen(typo)<5)continue;
      int min=0;
      int max=milon_count;
      while(max>=min){
         j=min+(max-min)/2;
         if (strcmp(typo,milon[j]) == 0) {
            correction=milon[j]+strlen(milon[j])+1;
            method=correction+strlen(correction)+1;
            location=method+strlen(method)+1;
            fprintf(fp3, "\nINSERT INTO `suspects` (`project`,`title`,`suspect`,`correction`,`type`,`location`,`status`,`fixer`) VALUES ('en.wikipedia','%s','%s','%s','%s','%s',0,'');", pagename, typo, correction, method, location);
            break;
         }
         if (strcmp(typo,milon[j])>0){min=j+1;}
         if (strcmp(typo,milon[j])<0){max=j-1;}
      }
   }

   fprintf(fp3,"\nFinally!");

   fclose(fp1);
   fclose(fp2);
   fclose(fp3);
   free(milon);
   return 0;
}

File: parsers/find-words-in-dump.py
# -*- coding: utf-8 -*-
import json
import re
import time;

with open('/Users/asafmalin/Documents/GitHub/systematic-editing/parsers/data/fr-variations.txt') as f:
   suspects = {}
   for line in f:
      (key, val) = line.split()
      suspects[key] = val

with open('/Users/asafmalin/Documents/GitHub/systematic-editing/parsers/data/frwiki-2022-08-29.txt') as f:
   existingWords = {}
   for line in f:
      (key, val) = line.split()
      existingWords[key] = int(val)

def decodeLine(line):
    line = line.replace('&quot;','"')
    line = line.replace('&lt;','<')
    line = line.replace('&gt;','>')
    line = line.replace('&amp;','&')
    line = line.replace('&nbsp;',' ')
    line = line.replace('&mdash;','—')
    return line

def countUnknownWords(history):
    historyStr = ' '.join(history)
    words = re.split(r"([\-{}\[,.\";\^~#\=&\)\|<>\?]*\s+[\-{},.\";\^~#\=&\)\|>\?]*)", historyStr)
    count = 0
    nonExisting = ''
    for index, currentword in enumerate(words):
        currentword = currentword.strip("'")
        hasNonLetter = False
        hasCapital = False
        # hasNonAscii = False
        #check if currentword contains capital letter
        for letter in currentword:
            if letter.isupper():
                hasCapital = True
            if not letter.isalpha() and letter != "'":
                hasNonLetter = True
            # if not letter.isascii():
            #     hasNonAscii = True
        if currentword and (not hasNonLetter and not hasCapital and currentword not in existingWords):
            nonExisting = nonExisting + ',' + currentword
            count += 1
    #if count > 1:
    #    print(nonExisting)
    return count

class isScanFlags:
    def __init__(self):
        self.mainflag = 1
        self.textflag = 1
    def newArticleReset(self):
        self.templateflag = 0
        self.externalflag = 0
        self.refflag = 0
        self.fileflag = 0
        self.quoteflag = 0
        self.quote2flag = 0
        self.quote3flag = 0
        self.commentflag = 0
        self.divflag = 0
        self.codeflag = 0
        self.poemflag = 0
        self.galleryflag = 0
        self.Galleryflag = 0
        self.sourceflag = 0
    def newLineReset(self):
        self.fileflag = 0
    
mainstr = '<ns>0</ns>'
endpage = '</page>'
#typos = []
#checkedWords = {}
title = ''
history = [''] * 9
f = open('/Users/asafmalin/Documents/GitHub/systematic-editing/parsers/data/fr'+str(time.time())+'results.txt', 'w')
from bz2 import BZ2File
with BZ2File('/Users/asafmalin/Documents/GitHub/systematic-editing/parsers/data/frwiki-20231120-pages-articles.xml.bz2','rb') as file:
    flags = isScanFlags()
    for line in file:
        flags.newLineReset()
        line = line.decode()
        line = decodeLine(line)
        for index, historyline in enumerate(history[::]):
            if not index:
                history[0] = line
                history[index + 1] = historyline
            elif index != len(history) - 1:
                history[index + 1] = historyline
        if '<title>' in line:
            title = line
            flags.newArticleReset()
        if '<text' in line:
            flags.textflag = 0
        if flags.textflag:
            continue
        if mainstr in line:
            flags.mainflag = 0
            flags.textflag = 1
        if endpage in line:
            flags.mainflag = 1
        if flags.mainflag:
            continue
        if line[0][0] == ':':
            continue
        previous = next_ = ''
        line = line.rstrip('\n')+' '
        objects = re.split(r"([\-{}\[,.\";\^~#\=&\)\|<>\?]*\s+[\-{},.\";\^~#\=&\)\|>\?]*)", line)
        l = len(objects)-2
        for index, word in enumerate(objects):
            if word == ' ':
                continue
            if '}}' in word and flags.templateflag > 0:
                flags.templateflag -= 1
            if '{{' in word:
                flags.templateflag += 1
            if '[[' in word and ':' in word:
                flags.fileflag = 1
            if ']' in word:
                flags.externalflag = 0
            if '[h' in word:
                flags.externalflag = 1
            if '<!--' in word:
                flags.commentflag = 1
            if '-->' in word: 
                flags.commentflag = 0
            if '<ref' in word:
                flags.refflag = 1
            if '/ref' in word:
                flags.refflag = 0
            if '<div' in word:
                flags.divflag = 1
            if '/div' in word:
                flags.divflag = 0
            if '<code' in word:
                flags.codeflag = 1
            if '/code' in word:
                flags.codeflag = 0    
            if '<poem' in word:
                flags.poemflag = 1
            if '/poem' in word:
                flags.poemflag = 0
            if '<gallery' in word:
                flags.galleryflag = 1
            if '/gallery' in word:
                flags.galleryflag = 0
            if '<Gallery' in word:
                flags.Galleryflag = 1
            if '/Gallery' in word:
                flags.Galleryflag = 0    
            if '<source' in word:
                flags.sourceflag = 1
            if '/source' in word:
                flags.sourceflag = 0    
            if (word.count('"') % 2) and not flags.quoteflag:
                flags.quoteflag = 1
                continue
            if (word.count('"') % 2) and flags.quoteflag:
                flags.quoteflag = 0
            if '\'\'' in word and not flags.quote2flag:
                flags.quote2flag = 1
                continue
            if '\'\'' in word and flags.quote2flag:
                flags.quote2flag = 0
            if '<block' in word:
                flags.quote3flag = 1
            if '</block' in word:
                flags.quote3flag = 0
            if flags.templateflag or flags.sourceflag or flags.Galleryflag or flags.galleryflag or flags.poemflag or flags.codeflag or flags.divflag or flags.quoteflag or flags.quote2flag or flags.quote3flag or flags.commentflag or flags.externalflag or flags.fileflag or flags.refflag:
                continue
            if not word.isalpha():
                continue
            if len(word) < 3:
                continue
            if not word.islower():
                continue
            if word not in suspects:
                continue
            if countUnknownWords(history) > 1:
                continue
            #if the word without last s in words
            #if word[0:len(word)-1:] and word[len(word)-1]=='s' and word[len(word)-2]!='s':
            #    continue
            #if word in engrem:
            #    continue
            #if word in checkedWords:
            #    continue
            #checkedWords[word] = True
            #if word not in dbvariations:
            #    continue
            #if word in typos:
            #    break
            
            row = suspects[word].split(",")
            variant = row[0]
            variationType = row[1]
            spot = int(row[2])
            
            #typos.append(word)
            before = ''
            after = ''
            trimmedTitle = title.replace('    <title>','').replace('</title>\n','').replace("'","\\'")
            word = word.replace("'","\\'")
            variant = variant.replace("'","\\'")            
            f.write(word+" INSERT INTO `suspects` (`project`,`title`,`suspect`,`correction`,`type`,`location`,`status`,`fixer`) VALUES ('en.wikipedia','"+trimmedTitle+"','"+word+"','"+variant+"','"+variationType+"','"+str(spot)+"',0,'');\n")
            #f.flush()
            break
f.close()
